# Task
Implement and evaluate an SVM model on the dataset provided in "/content/Housing.csv".

## Load the dataset

### Subtask:
Load the `Housing.csv` file into a pandas DataFrame.


**Reasoning**:
Import pandas and load the CSV file into a DataFrame, then display the head and info to understand the data.



In [1]:
import pandas as pd

df = pd.read_csv('/content/Housing.csv')
display(df.head())
display(df.info())

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


None

## Prepare the data

### Subtask:
Perform any necessary data cleaning, handle missing values, and encode categorical features.


**Reasoning**:
Identify and handle missing values, and encode categorical features in the DataFrame.



In [2]:
# Check for missing values
display(df.isnull().sum())

# Identify categorical columns
categorical_cols = df.select_dtypes(include='object').columns
display("Categorical columns to encode:", categorical_cols)

# Apply one-hot encoding to categorical columns
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# Display the first few rows of the modified DataFrame and check data types
display(df_encoded.head())
display(df_encoded.info())

price               0
area                0
bedrooms            0
bathrooms           0
stories             0
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     0
parking             0
prefarea            0
furnishingstatus    0
dtype: int64

'Categorical columns to encode:'

Index(['mainroad', 'guestroom', 'basement', 'hotwaterheating',
       'airconditioning', 'prefarea', 'furnishingstatus'],
      dtype='object')

price  area  bedrooms  bathrooms  stories  parking  mainroad_yes  \
0  13300000  7420         4          2        3        2          True   
1  12250000  8960         4          4        4        3          True   
2  12250000  9960         3          2        2        2          True   
3  12215000  7500         4          2        2        3          True   
4  11410000  7420         4          1        2        2          True   

   guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
0          False         False                False                 True   
1          False         False                False                 True   
2          False          True                False                False   
3          False          True                False                 True   
4           True          True                False                 True   

   prefarea_yes  furnishingstatus_semi-furnished  furnishingstatus_unfurnished  
0          True                            False                         False  
1         False                            False                         False  
2          True                             True                         False  
3          True                            False                         False  
4         False                            False                         False

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype
---  ------                           --------------  -----
 0   price                            545 non-null    int64
 1   area                             545 non-null    int64
 2   bedrooms                         545 non-null    int64
 3   bathrooms                        545 non-null    int64
 4   stories                          545 non-null    int64
 5   parking                          545 non-null    int64
 6   mainroad_yes                     545 non-null    bool 
 7   guestroom_yes                    545 non-null    bool 
 8   basement_yes                     545 non-null    bool 
 9   hotwaterheating_yes              545 non-null    bool 
 10  airconditioning_yes              545 non-null    bool 
 11  prefarea_yes                     545 non-null    bool 
 12  furnishingstatus_semi-furnished  545 non-null    b

None

## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the data into training and testing sets using `train_test_split`.



In [3]:
from sklearn.model_selection import train_test_split

X = df_encoded.drop('price', axis=1)
y = df_encoded['price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (436, 13)
Shape of X_test: (109, 13)
Shape of y_train: (436,)
Shape of y_test: (109,)


## Scale the features

### Subtask:
Scale the numerical features to ensure they have a similar range.


**Reasoning**:
Scale the numerical features in the training and testing sets using StandardScaler.



In [4]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Identify numerical columns (excluding the one-hot encoded boolean columns)
numerical_cols = ['area', 'bedrooms', 'bathrooms', 'stories', 'parking']

# Fit the scaler on the numerical features of the training data
X_train[numerical_cols] = scaler.fit_transform(X_train[numerical_cols])

# Transform the numerical features of both the training and testing data
X_test[numerical_cols] = scaler.transform(X_test[numerical_cols])

display(X_train.head())
display(X_test.head())

area  bedrooms  bathrooms   stories   parking  mainroad_yes  \
46   0.384168  0.055271   1.539173  2.587644  0.367957          True   
93   0.929181  0.055271   1.539173 -0.912499  2.709987          True   
335 -0.607755 -1.283514  -0.557950 -0.912499  1.538972          True   
412 -1.155492  0.055271  -0.557950  0.254215 -0.803059          True   
471 -0.637730  0.055271  -0.557950  0.254215 -0.803059          True   

     guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
46           False         False                False                 True   
93           False          True                False                 True   
335          False          True                False                 True   
412          False          True                False                False   
471          False         False                False                False   

     prefarea_yes  furnishingstatus_semi-furnished  \
46          False                            False   
93          False                             True   
335         False                            False   
412          True                            False   
471         False                            False   

     furnishingstatus_unfurnished  
46                          False  
93                          False  
335                         False  
412                          True  
471                          True

area  bedrooms  bathrooms   stories   parking  mainroad_yes  \
316  0.338750  1.394055   1.539173  0.254215  0.367957         False   
77   0.611257  0.055271   1.539173  1.420929 -0.803059          True   
360 -0.506019 -1.283514  -0.557950 -0.912499 -0.803059          True   
90  -0.070009  0.055271  -0.557950  0.254215 -0.803059          True   
493 -0.542353  0.055271  -0.557950 -0.912499 -0.803059          True   

     guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
316          False          True                False                False   
77           False         False                False                 True   
360          False         False                False                False   
90           False         False                False                 True   
493          False         False                False                False   

     prefarea_yes  furnishingstatus_semi-furnished  \
316         False                            False   
77           True                            False   
360         False                             True   
90          False                             True   
493         False                            False   

     furnishingstatus_unfurnished  
316                          True  
77                          False  
360                         False  
90                          False  
493                         False

## Train the svm model

### Subtask:
Train a Support Vector Machine (SVM) model on the training data.


**Reasoning**:
Import the SVR model, instantiate it, and train it on the scaled training data.



In [5]:
from sklearn.svm import SVR

# Instantiate an SVR object with default parameters
svr_model = SVR()

# Fit the SVR model to the scaled training data
svr_model.fit(X_train, y_train)

SVR()

## Evaluate the model

### Subtask:
Evaluate the performance of the trained SVM model on the testing data.


**Reasoning**:
Evaluate the performance of the trained SVM model on the testing data by calculating and printing the Mean Squared Error and R-squared score.



In [6]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the testing data
y_pred = svr_model.predict(X_test)

# Calculate and print Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse}")

# Calculate and print R-squared score
r2 = r2_score(y_test, y_pred)
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 5567929077615.067
R-squared (R2) Score: -0.10156329939722397


## Summary:

### Data Analysis Key Findings

*   The dataset contains 545 entries and 13 columns, including both numerical and categorical features, with no missing values.
*   Seven categorical columns were identified and successfully one-hot encoded.
*   The dataset was split into training (436 samples) and testing (109 samples) sets, with 'price' as the target variable.
*   Numerical features ('area', 'bedrooms', 'bathrooms', 'stories', 'parking') were scaled using `StandardScaler`.
*   An SVR (Support Vector Regression) model was trained on the scaled training data.
*   The trained SVR model achieved a Mean Squared Error (MSE) of approximately 5.57 x 10¹² and an R-squared (R2) score of approximately -0.10 on the test set.

### Insights or Next Steps

*   The current SVR model performs very poorly, indicated by the extremely high MSE and negative R2 score. This suggests that the default SVR parameters may not be suitable for this dataset, or SVM might not be the best model for this regression task.
*   Further steps should include hyperparameter tuning for the SVR model (e.g., using GridSearchCV) or exploring alternative regression models better suited for the dataset's characteristics.
